### PR TITLE
Feature/admin course management

### PR DESCRIPTION
**Pull request by tobias-wiese**
# Description

Implements issue #438 
Implements issue #439 
-> implements epic #200 

## Reason for this PR
- implement the epic

## Changes in this PR
- add course list and course form to the navbars of the admin
- adapt routing as list and form are now shared between admin and lecturer ("/all-courses")
- enable editing of all courses for the admin
- add a lecturer selection section to the course form that is visible just for admins

## Type of change (remove all that don't apply)
- New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Manually tested all use cases this could alter

**Test Configuration**:
- OS: Windows
- Browser: Firefox

- Frontend: (remove all that don't apply)
  - [x] Development build
- Backend: (remove all that don't apply)
  - [x] deployed development build

# Checklist: (remove all that don't apply)

- [x] I added corresponding E2E tests (especially for bugfixes)
- [x] I have performed a self-review of my own code
- [x] My changes generate no new linting warnings or console warnings
- [x] I have updated the CHANGELOG.md to include any changes made in this PR (add WIP to the top, if there is none already)

## Notes:
creating a course with an invalid lecturer Id "works" atm as we do not get a validation error. The validation error is implemented in the backend, but not yet merged. The error handling for that is implemented.

### CODE REVIEW

**Review by bastihav — CHANGES_REQUESTED** [review]
Feature is completely unusable on both chrome and firefox, did you forget to emit the lecturerId?

Whats the difference between the shared/ and common/ folder? Should we merge them?
I am not sure about the "lecturer found" text, at least the styling. I think a more subtle green checkmark or red cross + black text would be more suitible, maybe add a link to the profile of the lecturer too.

E2E tests are planned as soon as the test branch is merged, I guess?

**Comment by tobias-wiese** [comment]
> 
> 
> Feature is completely unusable on both chrome and firefox, did you forget to emit the lecturerId?

Yes, that was the problem. Should be fixed now.

> 
> Whats the difference between the shared/ and common/ folder? Should we merge them?

I added the "shared" folder for views, that are shown for multiple roles (here Lecturer and Admin), but not for all roles (as for example the profile pages). It should not be a problem to change merge the folders, I just thought this could be "cleaner".

> I am not sure about the "lecturer found" text, at least the styling. I think a more subtle green checkmark or red cross + black text would be more suitible, maybe add a link to the profile of the lecturer too.

I implement this like in the following screenshots
grafik
(This is the hover style for the link)
grafik

> 
> E2E tests are planned as soon as the test branch is merged, I guess?

Yes. I already added a very small test for the new navigation, but the functionality will be tested, as soon the test branch is merged.

File: src/views/shared/EditCreateCourseForm.vue
Comment (bastihav) [inline src/views/shared/EditCreateCourseForm.vue:90]:
```suggestion
    import { useStore } from "@/use/store/store";
```

did not test if this breaks things, but we should only access the store with getters anyway

**Review by bastihav — COMMENTED** [review]
will approve after e2e tests

File: tests/e2e/adminCourseManagement.spec.ts
Comment (bastihav) [inline tests/e2e/adminCourseManagement.spec.ts:16]:
```suggestion
            course = c as Course;
```

**Review by bastihav — APPROVED** [review]

**Comment by theovier** [comment]
I would prefer if the select menu for the lecturers (first image) would look more like the other selection menus we used (second image) just for the sake of consistency. [Chrome on MacOS used].
image
image

Besides that, it looks good and functional 👍

**Comment by tobias-wiese** [comment]
> 
> 
> I would prefer if the select menu for the lecturers (first image) would look more like the other selection menus we used (second image) just for the sake of consistency. [Chrome on MacOS used].
> image
> image
> 
> Besides that, it looks good and functional 👍

That's because it is a datalist connected to an input field, not a select field. I will try to style the datalist as well

**Comment by theovier** [comment]
> > I would prefer if the select menu for the lecturers (first image) would look more like the other selection menus we used (second image) just for the sake of consistency. [Chrome on MacOS used].
> > image
> > image
> > Besides that, it looks good and functional 👍
> 
> That's because it is a datalist connected to an input field, not a select field. I will try to style the datalist as well

Just curious, what is the reason that you used a datalist rather than a prefilled select field (like the countries)?

**Comment by tobias-wiese** [comment]
> 
> 
> > > I would prefer if the select menu for the lecturers (first image) would look more like the other selection menus we used (second image) just for the sake of consistency. [Chrome on MacOS used].
> > > image
> > > image
> > > Besides that, it looks good and functional 👍
> > 
> > 
> > That's because it is a datalist connected to an input field, not a select field. I will try to style the datalist as well
> 
> Just curious, what is the reason that you used a datalist rather than a prefilled select field (like for the languages)?

To enable a search. If we have like 100 lecturers, I want to filter the datalist with the input given in the text input field

**Comment by theovier** [comment]
> To enable a search. If we have like 100 lecturers, I want to filter the datalist with the input given in the text input field
You can have a search also with a custom select if I'm not mistaken. As I understood datalist they represent more of `suggestion` than actual choices. However, we require the user to select one of several predefined choices. However, I'm fine with it if the styling matches.

**Comment by tobias-wiese** [comment]
> 
> 
> > To enable a search. If we have like 100 lecturers, I want to filter the datalist with the input given in the text input field
> 
> You can have a search also with a custom select if I'm not mistaken. As I understood, datalist represent more of `suggestion` than actual choices. However, we require the user to select one of several predefined choices. However, I'm fine with it if the styling matches.

I do know that implementing such a select is possible. I decided to use the data list as it is a plain HTML5 solution. Otherwise I'd had to implement it using JS (with an input field and different divs for the options)  or use a plugin, but you know the problems with plugins 😄

**Comment by tobias-wiese** [comment]
> 
> 
> > > To enable a search. If we have like 100 lecturers, I want to filter the datalist with the input given in the text input field
> > 
> > 
> > You can have a search also with a custom select if I'm not mistaken. As I understood, datalist represent more of `suggestion` than actual choices. However, we require the user to select one of several predefined choices. However, I'm fine with it if the styling matches.
> 
> I do know that implementing such a select is possible. I decided to use the data list as it is a plain HTML5 solution. Otherwise I'd had to implement it using JS (with an input field and different divs for the options) or use a plugin, but you know the problems with plugins 😄

Now I implemented an own select component with an input and absolute divs that looks like our form selects (see screenshots). Styling the datalist was not possible as it is a browser specific component. @bastihav  You may want to look into the implementation again, as I changed and added many things
grafik

**Review by theovier — APPROVED** [review]
Looks nice, ty 😘

File: src/views/shared/EditCreateCourseForm.vue
Comment (bastihav) [inline src/views/shared/EditCreateCourseForm.vue:131]:
```suggestion
        setup(props: any, { emit }: any) {
```
not used

File: tests/e2e/adminCourseManagement.spec.ts
Comment (bastihav) [inline tests/e2e/adminCourseManagement.spec.ts:28]:
```suggestion
        cy.fixture("logins/lecturer.json").then((lecturer) => {
```

File: tests/e2e/adminCourseManagement.spec.ts
Comment (bastihav) [inline tests/e2e/adminCourseManagement.spec.ts:109]:
```suggestion
        cy.get("input[id='lecturerId']").clear().type(lecturerAuth.username);
```

File: tests/e2e/adminCourseManagement.spec.ts
Comment (bastihav) [inline tests/e2e/adminCourseManagement.spec.ts:110]:
```suggestion
        cy.get("div").contains(`(@${lecturerAuth.username})`).should("exist");
```

File: tests/e2e/adminCourseManagement.spec.ts
Comment (bastihav) [inline tests/e2e/adminCourseManagement.spec.ts:120]:
```suggestion
        cy.get("div").contains(`(@${lecturerAuth.username})`).click();
        cy.get("input[id='lecturerId']").should("have.value", lecturerAuth.username);
```

**Review by bastihav — CHANGES_REQUESTED** [review]
did not test any of the changes, if the tests pass, feel free to merge

**Pull request merged**
